### PR TITLE
make implicit cstring conversions explicit

### PR DESCRIPTION
**Pull request by ee7**
The Nim manual [says](https://github.com/nim-lang/Nim/blob/772ed5eff296/doc/manual.rst#cstring-type) that an implicit conversion to cstring will
eventually not be allowed:

> A Nim `string` is implicitly convertible to `cstring` for convenience.
>
> [...]
>
> Even though the conversion is implicit, it is not *safe*: The garbage collector
> does not consider a `cstring` to be a root and may collect the underlying
> memory. For this reason, the implicit conversion will be removed in future
> releases of the Nim compiler. Certain idioms like conversion of a `const` string
> to `cstring` are safe and will remain to be allowed.

And [from Nim 1.6.0](https://github.com/nim-lang/Nim/blob/772ed5eff296/changelogs/changelog_1_6_0.md#type-system), such a conversion triggers a warning:

> A dangerous implicit conversion to `cstring` now triggers a `[CStringConv]` warning.
> This warning will become an error in future versions! Use an explicit conversion
> like `cstring(x)` in order to silence the warning.

However, some modules in this repo produced such a warning. For example,
before this commit, compiling `parsejson.nim` would produce:

> /foo/Nim/lib/pure/parsejson.nim(221, 37) Warning: implicit conversion to 'cstring' from a non-const location: my.buf; this will become a compile time error in the future [CStringConv]
> /foo/Nim/lib/pure/parsejson.nim(231, 39) Warning: implicit conversion to 'cstring' from a non-const location: my.buf; this will become a compile time error in the future [CStringConv]

This commit resolves the most visible `CStringConv` warnings, making the
cstring conversions explicit.

---

I generated the list of warnings by running `nim check` on every `.nim` file in this repo (excluding those in the `tests` directory) on Linux, macOS, and Windows. So this PR currently won't resolve `CStringConv` warnings that are only triggered on e.g. BSD.

### CODE REVIEW

**Comment by ee7** [comment]
I'll be able to fix at least two more (for JS only, and BSD only) - please don't merge quite yet.

**Comment by Varriount** [comment]
Once the changes are complete, move the PR out of *draft* status so that it can be reviewed.

**Comment by ee7** [comment]
Forgot about this PR. Sorry for the delay.

I resolved a couple more warnings, and rebased on `devel`. 

CI failure due to [one job exceeding the maximum execution time of 60 minutes](https://github.com/nim-lang/Nim/actions/runs/2855403824):

> **Annotations**
> 2 errors
>
> **macos-11 (batch: 2_3)**
> The job running on runner GitHub Actions 3 has exceeded the maximum execution time of 60 minutes.
>
> **macos-11 (batch: 2_3)**
> The operation was canceled.

 I am not able to restart the failing job.

**Comment by ringabout** [comment]
> I am not able to restart the failing job.

done

File: lib/pure/logging.nim [hunk 1]
```diff
@@ -367,7 +367,7 @@ method log*(logger: ConsoleLogger, level: Level, args: varargs[string, `$`]) =
   if level >= logging.level and level >= logger.levelThreshold:
     let ln = substituteLog(logger.fmtStr, level, args)
     when defined(js):
-      let cln: cstring = ln
+      let cln = ln.cstring
```
Comment (ee7) [inline lib/pure/logging.nim:370]: This is the first change that was made after the `merge_when_passes_CI` label was added.

File: lib/pure/osproc.nim [hunk 2]
```diff
@@ -1174,14 +1174,14 @@ elif not defined(useNimRtl):
...
           # uClibc environment (OpenWrt included) doesn't have the full execvpe
           let exe = findExe(data.sysCommand)
-          discard execve(exe, data.sysArgs, data.sysEnv)
+          discard execve(exe.cstring, data.sysArgs, data.sysEnv)
         else:
           # MacOSX doesn't have execvpe, so we need workaround.
           # On MacOSX we can arrive here only from fork, so this is safe:
           environ = data.sysEnv
-          discard execvp(data.sysCommand, data.sysArgs)
+          discard execvp(data.sysCommand.cstring, data.sysArgs)
```
Comment (ee7) [inline lib/pure/osproc.nim:1182]: This is the second of the two changes that were made after the `merge_when_passes_CI` label was added.

**Review by ee7 — COMMENTED** [review]
In case it helps a reviewer, I've highlighted the two changes that I found after the `merge_when_passes_CI` label was applied.

**Pull request merged**